File: CompetitiveProgramming/DWITE/DWITE '08 R3 #2 - Wordcount++/2021-07-12/WordCountPPRegEx.cpp
/*
////////////////////////////////////////////////////////////////////////
//////////////////////////// SYNTAX SUMMARY ////////////////////////////
1. C++ Regular Expression: Count Matches
    https://stackoverflow.com/questions/8283735/count-number-of-matches
    https://www.ibm.com/docs/en/zos/2.3.0?topic=types-sregex-iterator

    // ptrdiff_t is a signed integer storing the result of two pointer subtraction.
    ptrdiff_t const matches
    (
        // std::distance() finds the number of elements between two iterators.
        distance
        (
            // sregex_iterator() is a typedef over regex_iterator for iterators of type string::const_iterator.
            // regex_iterator() iterates through matches in string.
            sregex_iterator(s.begin(), s.end(), pattern),
            sregex_iterator()
        )
    );

///////////////////////////////////////////////////////////////////////////////////////
//////////////////////////// DWITE '08 R3 #2 - Wordcount++ ////////////////////////////
Given five lines of input, find the number of words that are at least four characters in length.
A word is any consecutive sequence of alpha (uppercase and lowercase) characters.

Sample Input
DWITE question number two: Wordcount++
Note how "two:" does not count.
For simplicity - "don't" is two words, split by apostrophe.

That was a blank line above.

Sample Output
4
3
4
0
4

////////////////////////////////////////////////////////////////////////
//////////////////////////// SOLUTION LOGIC ////////////////////////////
Match words with regular expression string. Match by 2 rules:
    1. alpha.
    2. at least 4 consecutive alpha characters.
*/

/**
 * Program to count the number of valid words per line.
 * @author Vincent Zhang
 * @since 12 July 2021
 */
#include <iostream>
#include <regex>
using namespace std;

/**
 * Main function to solve the problem.
 * @return 0
 */
int main()
{
    // Fast input.
    cin.sync_with_stdio(0);
    cin.tie();

    // Regex pattern.
    regex const pattern("[a-zA-Z]{4,}");

    // Loop through the 5 lines.
    for(int i=0; i<5; i++)
    {
        string s; getline(cin, s);

        // ptrdiff_t is a signed integer storing the result of two pointer subtraction.
        ptrdiff_t const matches
        (
            // std::distance() finds the number of elements between two iterators.
            distance
            (
                // sregex_iterator() is a typedef over regex_iterator for iterators of type string::const_iterator.
                // regex_iterator() iterates through matches in string.
                sregex_iterator(s.begin(), s.end(), pattern),
                sregex_iterator()
            )
        );
        // Output.
        cout << matches << '\n';
    }
}

File: CompetitiveProgramming/DWITE/DWITE '08 R3 #2 - Wordcount++/2021-07-12/WordCountPPRegEx.py
"""
######################################################################################
############################ PYTHON REGEX & REGEX PATTERN ############################
1. Get List of Matches
    import re
    newList = re.findall("REGEX PATTERN", STRING TO SEARCH IN)

2. Match Alpha
    "[a-zA-Z]"

3. Match by Minimum Occurance
    "{n,}" <-- previous match occures at least n times.

4. Match Alpha & Minimum Occurance
    "[a-zA-Z]{4,}" match words where alpha characters consecutively appear at least 4 times.

#######################################################################################
############################ DWITE '08 R3 #2 - Wordcount++ ############################
Count number of words with length > 3 for each of the 5 lines of input.

Sample Input
DWITE question number two: Wordcount++
Note how "two:" does not count.
For simplicity - "don't" is two words, split by apostrophe.

That was a blank line above.

Sample Output
4
3
4
0
4

########################################################################
############################ SOLUTION LOGIC ############################
Match words with regular expression. Two regex conditions must be satisfied:
    1. is alpha.
    2. at least 4 consecutive alpha characters.

#########################################################################
############################ MODULE COMMENTS ############################
@author Vincent Zhang
@since 12 July 2021
@context Learning Python 3
"""

# Use sys for fast input. Use re for regular expression.
import sys, re

# For each of the 5 lines of input, print the list of valid words that regex matches.
for _ in range(5): print(len(re.findall("[a-zA-Z]{4,}", sys.stdin.readline())))